Add optional structured logs of RPC related events

rem1-dev opened the pull request:
Adding RPC related event logs that will be used in Blockstream Explorer project. The logs are structured as json object. This optional logging may be used in an enterprise/internal scenario for billing purposes. By default it's turned off. 

The logs are controlled with a new `--rpc-logging` command line option. Command line option possible values, and example structured logs they produce, are as following:

- `--rpc-logging full`  results in logs:

```
ELECTRUM-RPC-LOGGER: {"event":"connection established","source":{"ip":"127.0.0.1","port":59877}}
ELECTRUM-RPC-LOGGER: {"event":"rpc request","id":18,"method":"blockchain.scripthash.get_history","params":["6b63eef944d982701eb2d0dbb8ee900f42d8e79fe3d1ea473602c0edc87c34f6"],"source":{"ip":"127.0.0.1","port":59877}}
ELECTRUM-RPC-LOGGER: {"duration_µs":114,"event":"rpc response","id":18,"method":"blockchain.scripthash.get_history","payload_size":37,"source":{"ip":"127.0.0.1","port":59877}}
ELECTRUM-RPC-LOGGER: {"event":"connection closed","source":{"ip":"127.0.0.1","port":59877}}
```

- `--rpc-logging no-params` excludes RPC request params to be logged, results in logs like:

```
ELECTRUM-RPC-LOGGER: {"event":"connection established","source":{"ip":"127.0.0.1","port":59886}}
ELECTRUM-RPC-LOGGER: {"event":"rpc request","id":18,"method":"blockchain.scripthash.get_history","params":null,"source":{"ip":"127.0.0.1","port":59886}}
ELECTRUM-RPC-LOGGER: {"duration_µs":78,"event":"rpc response","id":18,"method":"blockchain.scripthash.get_history","payload_size":37,"source":{"ip":"127.0.0.1","port":59886}}
ELECTRUM-RPC-LOGGER: {"event":"connection closed","source":{"ip":"127.0.0.1","port":59886}}```

rem1-dev committed:
Add optional structured logs of RPC related events

shesek commented on src/config.rs:
```diff
@@ -39,6 +39,7 @@ pub struct Config {
     pub utxos_limit: usize,
     pub electrum_txs_limit: usize,
     pub electrum_banner: String,
+    pub rpc_logging: Option<RpcLogging>,
```
The CLI option name should be prefixed with `electrum` like the other electrum-related options.

shesek replied:
It could be `electrum_logging` without `rpc` if you prefer to keep things short

rem1-dev replied:
I changed this name in newest commit

shesek commented on src/electrum/server.rs:
```diff
@@ -490,6 +493,27 @@ impl Connection {
         Ok(result)
     }
 
+    fn log_rpc_event(&self, entries: &Vec<(&str, Value)>) {
```
This should ideally avoid preparing and allocating the log `entries` when logging is turned off to avoid the unnecessary performance hit.

I would consider using a macro to make `entries` evaluate only when logging is turned on. Alternatively this can be lazily evaluated by passing a thunk closure that computes `entries`, or simply by wrapping the outer calls in an `if`.

rem1-dev replied:
I've used macro that creates log entries only if `--electrum-rpc-logging` flag is set (see latest [commit](bfa5b9f84e8fcbb77da9f7f6d8b3720bfb683250))

rem1-dev committed:
Allocating RPC logs entries only if --electrum-rpc-logging is set

rem1-dev committed:
CI fix

shesek commented on src/electrum/server.rs:
```diff
@@ -490,6 +501,27 @@ impl Connection {
     }
 
+    fn log_rpc_event(&self, entries: &Vec<(&str, Value)>) {
+        if let Some(_) = self.rpc_logging {
```
Since this is now only ever called from the macro, maybe drop the `if` here?

rem1-dev replied:
You're right. I removed that.

rem1-dev committed:
Fixing review comment

shesek commented on src/electrum/server.rs:
```diff
@@ -490,6 +501,25 @@ impl Connection {
+    fn log_rpc_event(&self, entries: &Vec<(&str, Value)>) {
+        let mut log = json!({});
+
+        if let Some(log_map) = log.as_object_mut() {
```
This should never fail, I would consider using `unwrap()`

shesek replied:
Or alternatively construct `log` as a HashMap first, then convert to a json. This also somewhat simplifies the code:

```
fn log_rpc_event(&self, entries: Vec<(&str, Value)>) {
    let mut log = entries.into_iter().collect::<HashMap<_, _>>();
    log.insert(
        "source".into(),
        json!({
            "ip": self.addr.ip().to_string(),
            "port": self.addr.port(),
        }),
    );
    info!("{}", json!(log));
}
```

Note that this requires passing ownership over `entries` from the macro, which also reduces copying.

shesek replied:
Or perhaps even pass a `Value` `json!` map from the caller side to begin with and only inject `source` into it. This should simplify the code the most, as the `json!` syntax is terser compared to the `Vec` construction.

rem1-dev replied:
Suggested change applied.

shesek replied:
What my latest comment suggested was to pass a `serde::Value` map, not an `HashMap`:

```rust
fn log_rpc_event(&self, mut log: Value) {
    log.as_object_mut()
        .expect("log entry to be an object")
        .insert(
            "source".into(),
            json!({
                "ip": self.addr.ip().to_string(),
                "port": self.addr.port(),
            }),
        );
    println!("ELECTRUM-RPC-LOGGER:{}", log);
}
```

Then we could use the `json!()` syntax to construct the log entry, like so:
```rust
conditionally_log_rpc_event!(self, json!({
    "event": "rpc response",
    "payload_size": line.as_bytes().len(),
    "method": method_info,
}));
```

rem1-dev replied:
Corrected, now it looks much more compact, thanks for comment!

shesek commented on src/electrum/server.rs:
```diff
@@ -490,6 +501,25 @@ impl Connection {
+            );
+        }
+
+        info!("{}", log);
```
Maybe add a prefix to the log message so they're easier to locate? Something like `ELECTRUM-RPC-LOGGER:{...json...}`

shesek replied:
Also, I would consider printing this directly to `STDOUT` rather than going through the log macro, so that it doesn't depend on the configured log level and doesn't have the `INFO` prefix (which seems unnecessary, and may contain ANSI coloring sequences)

rem1-dev replied:
Suggested change applied. Now the log looks like:

````
INFO - NetworkInfo { version: 240100, subversion: "/Satoshi:24.1.0/", relayfee: 1e-5 }
INFO - BlockchainInfo { chain: "regtest", blocks: 500, headers: 500, bestblockhash: "53d198810c3a77721bdec389a6ac85c902b35069330099faea80d82456bed606", pruned: false, verificationprogress: 1.0, initialblockdownload: Some(false) }
INFO - Electrum RPC server running on 127.0.0.1:50001
INFO - REST server running on 127.0.0.1:3000
INFO - [127.0.0.1:57226] connected peer
ELECTRUM-RPC-LOGGER: {"event":"connection established","source":{"ip":"127.0.0.1","port":57226}}
ELECTRUM-RPC-LOGGER: {"event":"rpc request","method":"blockchain.scripthash.get_history","source":{"ip":"127.0.0.1","port":57226}}
ELECTRUM-RPC-LOGGER: {"event":"rpc response","method":"blockchain.scripthash.get_history","payload_size":38,"source":{"ip":"127.0.0.1","port":57226}}
ELECTRUM-RPC-LOGGER: {"event":"connection closed","source":{"ip":"127.0.0.1","port":57226}}
INFO - [127.0.0.1:57226] disconnected peer
```

rem1-dev committed:
log_rpc_event refactoring

rem1-dev committed:
Refactoring

shesek commented on src/electrum/server.rs:
```diff
@@ -490,6 +501,17 @@ impl Connection {
+                "port": self.addr.port(),
+            }),
+        );
+        println!("ELECTRUM-RPC-LOGGER: {}", json!(log));
```
The `json!()` here isn't necessary since `log` already is a json Value

All LGTM other than that. thanks!

rem1-dev replied:
Right, corrected.

rem1-dev committed:
Removing unnecessary json! call

shesek commented:
I was thinking that it might be useful to log two extra things - the jsonrpc `id` field so that the `rpc request` and `rpc response` entries can be matched to each-other, and the time it took to process the request.

I'm not sure whether these are necessary or not. If not, we can go ahead and merge.

rem1-dev commented:
Yes, it's good idea to add id into req and resp logs. Did you mean id that we can get from cmd variable: `cmd.get("id")`? I added that.

rem1-dev committed:
Adding json id to log

shesek committed:
Refactor: handle rpc replies inside the `match` arm

This makes `method_into` unnecessary and allows using the `id` variable
instead of reading `cmd` again. This also makes it easier to log time
durations (added in a followup commit).

shesek committed:
Some minor formatting and refactoring changes

shesek committed:
Log the time duration for processing RPC requests

shesek commented:
I made some minor formatting/refactoring suggestions and added the time duration for processing requests [in a branch](https://github.com/shesek/electrs/commits/rpc-logging/). What do you think?

rem1-dev commented:
It looks good to me. I also updated  PR description to be more verbose on why we're adding it.

shesek commented:
Great :) Can you apply the changes on your branch and rebase?

rem1-dev committed:
Change time duration for processing RPC requests log to microseconds

rem1-dev commented:
I applied your changes + I changed duration timestamp field to microseconds because when I tested it I never got duration > 1s and that field was always 0. Re rebase: not sure what should I rebase on. Branch is up-to date with new-index

shesek commented:
Thanks!

I accidentally merged without squashing, looking into fixing this
